Add migration versioning via Migration subclasses

matthewd opened the pull request:
This PR implements the migration versioning infrastructure, and restores the previous default on `timestamps null:` (when in 4.2-compatibility mode).

artofhuman commented on activerecord/lib/active_record/migration.rb:
```diff
@@ -414,6 +414,28 @@ def initialize
+    def self.inherited(subclass) # :nodoc:
+      super
+      if subclass.superclass == Migration
+        subclass.send :include, Compatibility::Legacy
```
mb use ? `subclass.include Compatibility::Legacy`

tenderlove commented:
LGTM.  I guess the main things to point out is the difference between upgrade strategies, namely [this](https://github.com/rails/rails/pull/21037/files#diff-b69cba6d94e5addb3b99326328edc25dR29) vs [this](https://github.com/matthewd/rails/blob/migration-version/activerecord/lib/active_record/migration/compatibility.rb#L33).  I think the strategy in this PR is easier.  But, FWIW, both PRs generate a superclass, so we can switch to the strategy pattern in #21037 at pretty much any point if we need to.

The other thing is that this PR doesn't support deriving the version from the file name.  IMO the syntax proposed here is pretty unoffensive, so I'm not sure if we still want / need to derive this from the file name.

I think the main hurdle here is that we need to figure out if the filename stuff is still required.

/cc @dhh @jeremy

dhh commented:
I'm happy enough with the superclass having the version identifier. Don't need the filename to include it too, imo. This is actually a pretty elegant solution.

matthewd committed:
Add migration versioning via Migration subclasses

matthewd committed:
In 4.2 migrations, `timestamps` defaulted to `null: true`

.. it also showed a deprecation warning, but we obviously needn't retain
that.

matthewd committed:
Use a deliberately-invalid migration version in all doc examples

If we use a real version, at best that'll be an onerous update required
for each release; at worst, it will encourage users to write new
migrations against an older version than they're using.

The other option would be to leave these bare, without any version
specifier. But as that's just a variant spelling of "4.2", it would seem
to raise the same concerns as above.

matthewd committed:
Ensure generated migrations include a version number

matthewd committed:
Internal test migrations use the private 'Current' version

Apart from specific versioning support, our tests should focus on the
behaviour of whatever version they're accompanying, regardless of when
they were written.

Application code should *not* do this.

matthewd committed:
Find the delegate, even in a deeper inheritance tree

matthewd committed:
Schema uses current migration API

matthewd committed:
Use a real migration version number in docs

Even though this means more things to change when we bump after a
release, it's more important that our examples are directly copyable.

jeremy commented:
:metal:
